File: day2/ruby/puzzle1.rb
data = File.read File.join(__dir__, '/../input.txt')


doubles, triples = data.split("\n").inject([0,0]) do |(doubles, triples), box|
  letters = box.chars
  processed = {}
  has_triple = false
  has_double = false

  letters.each do |char|
    unless processed.has_key? char
      letter_count = letters.count { |l| l == char }

      if letter_count == 3
        has_triple = true
      elsif letter_count == 2
        has_double = true
      end
    end
  end

  doubles += 1 if has_double
  triples += 1 if has_triple
  [doubles, triples]
end

puts "Doubles: #{doubles}"
puts "Triples: #{triples}"
puts "Checksum: #{doubles * triples}"

File: day2/python/puzzle1.py
import os

dir_path = os.path.dirname(os.path.realpath(__file__))
data = []

with open(dir_path + "/../input.txt", "r") as datafile:
    for line in datafile:
        data.append(line)

doubles = 0
triples = 0

for box in data:
    processed = {}
    has_triple = False
    has_double = False
    for char in box:
        if char not in processed.keys():
            processed[char] = box.count(char)

            if processed[char] == 3:
                has_triple = True
            elif processed[char] == 2:
                has_double = True
    if has_triple:
        triples += 1
    if has_double:
        doubles += 1

print(f"Doubles: {doubles}")
print(f"Triples: {triples}")
print(f"Checksum: {doubles * triples}")
